### PR TITLE
Price improvement tests

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
Fixes https://github.com/cowprotocol/services/issues/2479

### CODE REVIEW

**Comment by squadgazzz** [comment]
Still not quite sure this PR should replace all of [unit tests](https://github.com/cowprotocol/services/blob/dd233e23082f11509f1b1902004ac30b4b61e3bf/crates/driver/src/domain/competition/solution/fee.rs#L239-L353)

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 1]
```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
 
     protocol_fee_test_case(test_case).await;
 }
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_buy_out_market_order() {
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:194]:
```suggestion
async fn price_improvement_fee_buy_out_of_market_order() {
```

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 2]
```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
...
+        // executed buy amount should match quote buy amount
+        executed_buy_amount: quote_buy_amount.into(),
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_sell_out_market_order() {
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:226]:
```suggestion
async fn price_improvement_fee_sell_out_of_market_order() {
```

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 3]
```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
...
+        },
+    };
+    let test_case = TestCase {
+        order_side: order::Side::Buy,
+        fee_policy,
+        order_sell_amount: 50000000000000000000u128.into(),
+        solver_fee: Some(10000000000000000000u128.into()),
+        quote_sell_amount: quote_sell_amount.into(),
+        quote_buy_amount: quote_buy_amount.into(),
+        executed: quote_buy_amount.into(),
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:214]: How can we charge any fee here? The quote amount is equal to the executed amount so there is no price improvement. Doesn't this test framework now need a way of differentiating between the quoted and executed price? Also, are we even using an executed amount that is lower than the full amount in any test case (ie are we testing partial fills somewhere)

Overall, I still find these test cases a bit hard to parse (imagine someone trying to debug a failing test case and trying to reproduce the assumptions and math that is going on here).

I don't think we necessarily have to find the nicest abstraction that leads to the least code (e.g. I think we could have a different TestCase setup for price improvement fee vs. surplus fee if one is structurally more complex), but let's please make those tests easier to comprehend.

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 4]
```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
...
+            buy_amount: quote_buy_amount.into(),
+            fee: 1000000000000000000u128.into(),
+        },
+    };
+    let order_sell_amount = 50000000000000000000u128;
+    let test_case = TestCase {
+        order_side: order::Side::Sell,
+        fee_policy,
+        order_sell_amount: order_sell_amount.into(),
+        solver_fee: Some(10000000000000000000u128.into()),
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:244]: nit: I think it would be clearer to call this `network_fee` in this context.

I do see now that the price improvement comes from the network fee being different than the quoted fee. This is very clever but not something I'd expect someone maintaining this code in 3 month to just guess without context, so please let's help your future colleague to understand how this test work more intuitively.

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 5]
```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
...
+        quote_sell_amount: quote_sell_amount.into(),
+        quote_buy_amount: quote_buy_amount.into(),
+        executed: 40000000000000000000u128.into(),
+        // executed sell amount should match order sell amount
+        executed_sell_amount: order_sell_amount.into(),
+        executed_buy_amount: 44558823529411764706u128.into(),
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:254]: Partially fillable orders are missing.

Also can we add a test that shows shifting surplus into network fees doesn't change the protocol fee charged? E.g. the following two executions pay the same fee (assuming same underlying orders & quotes)

1. Sell 0.1 ETH receive 3500 DAI, pay 0.9 ETH in network fees
2. Sell 1 ETH receive 3500 DAI, pay 0 in network fees

File: crates/driver/src/tests/setup/mod.rs [hunk 6]
```diff
@@ -72,13 +72,28 @@ pub enum Score {
     RiskAdjusted { success_probability: f64 },
 }
 
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, serde::Serialize)]
+#[serde(rename_all = "camelCase", tag = "kind")]
+pub struct PriceImprovementQuote {
+    pub buy_amount: eth::U256,
+    pub sell_amount: eth::U256,
+    pub fee: eth::U256,
```
Comment (fleupold) [inline crates/driver/src/tests/setup/mod.rs:81]: if we define this for this test, let's also call this fee here network_fee (for clarity)

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/driver/src/tests/cases/protocol_fees.rs]: Before refactoring, now I understand that I am confused about a condition when an order is inside our outside market price.
Can out of market orders be defined as `(order.sell + order.fee) * quote.buy < (quote.sell + quote.fee) * order.buy`?

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 5 above)
Comment (squadgazzz) [reply crates/driver/src/tests/cases/protocol_fees.rs:249]: Partially fillable orders require some further refactoring. I will create a separate issue for this.

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 3 above)
Comment (fleupold) [reply crates/driver/src/tests/cases/protocol_fees.rs]: I believe so, yes (looking at @sunce86 to confirm)

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 7]
```diff
@@ -187,3 +188,121 @@ async fn volume_protocol_fee_sell_order() {
...
+            network_fee: 1000000000000000000u128.into(),
+        },
+    };
+    let order_sell_amount = 50000000000000000000u128.into();
+    let order_buy_amount = 40000000000000000000u128.into();
+    let test_case = TestCase {
+        order_side: order::Side::Buy,
+        fee_policy,
+        order_sell_amount,
+        network_fee: Some(2000000000000000000u128.into()),
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:210]: This will likely be easier to see with #2472, but can you comment in code how the price improvement is achieved (realized network fee lower than quote, leading to X price improvement). This will help understand the expected executed sell amount (compared to the quoted one), but maybe you can add a comment for this expectation as well.

Should be done for each test.

**Review by fleupold — APPROVED** [review]
Can we capture the follow up item of adding tests for partial fills (I believe this might also be missing for the surplus fee policy)?

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 7 above)
Comment (squadgazzz) [reply crates/driver/src/tests/cases/protocol_fees.rs]: Started writing the comments and found out that the results look pretty weird. Found [this code](https://github.com/cowprotocol/services/blob/8f8ef93bf6bdf27a77995f1897559010dbd8d004/crates/driver/src/tests/setup/blockchain.rs#L97-L110) which applies [a default surplus](https://github.com/cowprotocol/services/blob/8f8ef93bf6bdf27a77995f1897559010dbd8d004/crates/driver/src/tests/cases/mod.rs#L24) factor. Not sure if it needs to be eliminated for the price improvement tests.

**Review by Mateo-mro — APPROVED** [review]

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 7 above)
Comment (fleupold) [reply crates/driver/src/tests/cases/protocol_fees.rs]: Yes, either we need to set it in the test case or make sure it's set to 1. I think your trick with reducing the fee is good enough and allows us to more precisely reason about absolute surplus values (rather than a factor), so I'd slightly prefer setting the factor to 1 (but no strong feelings, as longs as it's clear from the main test how the calculation is carried out)

File: crates/driver/src/tests/setup/blockchain.rs [hunk 8]
```diff
@@ -97,14 +97,34 @@ impl QuotedOrder {
...
+            order::Side::Sell => {
+                let fee = match &self.order.fee_policy {
+                    FeePolicy::PriceImprovement {
+                        factor: _,
+                        max_volume_factor: _,
+                        quote,
+                    } => quote.network_fee,
+                    _ => eth::U256::zero(),
+                };
+                (self.buy * (self.sell - fee) / self.sell) / self.order.surplus_factor
```
Comment (squadgazzz) [inline crates/driver/src/tests/setup/blockchain.rs:109]: Is it valid?

File: crates/driver/src/tests/setup/blockchain.rs [hunk 9]
```diff
@@ -97,14 +97,34 @@ impl QuotedOrder {
...
+            order::Side::Buy => {
+                let fee = match &self.order.fee_policy {
+                    FeePolicy::PriceImprovement {
+                        factor: _,
+                        max_volume_factor: _,
+                        quote,
+                    } => quote.network_fee,
+                    _ => eth::U256::zero(),
+                };
+                (self.sell + fee) * self.order.surplus_factor
```
Comment (squadgazzz) [inline crates/driver/src/tests/setup/blockchain.rs:126]: Is it valid?

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 10]
```diff
@@ -59,6 +76,41 @@ async fn protocol_fee_test_case(test_case: TestCase) {
     test.solve().await.ok().orders(&[order]);
 }
 
+async fn price_improvement_fee_test_case(test_case: PriceImprovementTestCase) {
+    let test_name = format!("Protocol Fee: {:?} PriceImprovement", test_case.order_side);
+    let liquidity_quote = ab_liquidity_quote()
+        .sell_amount(test_case.order_sell_amount)
+        .buy_amount(test_case.order_buy_amount);
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:83]: Why is that set from the order amounts and not from the quote?

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 11]
```diff
@@ -136,55 +192,173 @@ async fn surplus_protocol_fee_sell_order_capped() {
...
+        network_fee: quote_network_fee,
+    };
+    let test_case = PriceImprovementTestCase {
+        order_side: order::Side::Buy,
+        policy_factor: 0.5,
+        policy_max_volume_factor: 1.0,
+        quote,
+        order_sell_amount,
+        order_buy_amount,
+        network_fee: Some(1.ether().into_wei()),
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:272]: You are still not saying anywhere that this is the little detail that leads to price improvement. This is way to hidden to leave this uncommented.

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 12]
```diff
@@ -136,55 +192,173 @@ async fn surplus_protocol_fee_sell_order_capped() {
...
+        order_side: order::Side::Sell,
+        policy_factor: 0.5,
+        policy_max_volume_factor: 1.0,
+        quote,
+        order_sell_amount,
+        order_buy_amount,
+        network_fee: Some(network_fee),
+        executed: order_sell_amount - network_fee,
+        executed_sell_amount: order_sell_amount,
+        // todo: how to prove the value?
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:303]: Well how is it calculated? Comment with the math that leads to this value.

Let's maybe start by "where is the price improvement coming from in this test"? I see sell amount and buy amounts and network fees are equal for quote and orders. Unclear why there is any price improvement fee here?

File: crates/driver/src/tests/setup/blockchain.rs (hunk 8 above)
Comment (fleupold) [reply crates/driver/src/tests/setup/blockchain.rs:109]: I think we should always incorporate some network fee here (and not have an if branch just for PriceImprovement fees. That feels way too hacky.

File: crates/driver/src/tests/setup/blockchain.rs (hunk 9 above)
Comment (fleupold) [reply crates/driver/src/tests/setup/blockchain.rs:126]: Same

**Review by fleupold — CHANGES_REQUESTED** [review]
I still find the whole setup very hacky and cannot see myself maintaining these tests.

Can we limit our price improvement for now to be only based on the difference in estimted network fee and actually network fee?

Then, a test case just needs a quote (which determines the order amounts and the expected fee), an actual fee amount (which determines the price improvement), a side and the expected amounts.

We shouldn't create if branches deep inside the test setup to cater to one specific test case but instead rethink the setup to make it work with all fee models.

**Comment by fleupold** [comment]
> Can we limit our price improvement for now to be only based on the difference in estimated network fee and actually network fee?

Actually maybe this wasn't a good idea to begin with (assuming it's the reason for the conditional refactoring). In this case, you can also work with explicit order and quote amounts and just assume 0 network everywhere.

**Comment by fleupold** [comment]
@squadgazzz I created https://github.com/cowprotocol/services/pull/2510 as a suggestion on how I would implement these tests. Can you take a look, tell me what you think about this approach and if it works and if convert the remaining tests?

**Comment by squadgazzz** [comment]
#2510
